domain/service/employees.go

```diff
@@ -0,0 +1,187 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/Shelex/grpc-go-demo/domain/entities"
+	"github.com/Shelex/grpc-go-demo/proto"
+	"github.com/Shelex/grpc-go-demo/storage"
+	"github.com/Shelex/grpc-go-demo/storage/documents"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type employeeService struct {
+	repository storage.Storage
+	documents  documents.FileStorage
+	grpc       *grpc.Server
+}
+
+func (e *employeeService) Employees(req *proto.GetAllRequest, stream proto.EmployeeService_EmployeesServer) error {
+	employees, err := e.repository.GetAll()
+	if err != nil {
+		return fmt.Errorf("failed to get employees: %w", err)
+	}
+	for _, emp := range employees {
+		if err := stream.Send(&proto.EmployeeResponse{Employee: entities.EmployeeFromStorageToProto(emp)}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *employeeService) EmployeeByID(ctx context.Context, req *proto.ByIDRequest) (*proto.EmployeeResponse, error) {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		log.Printf("metadata received: %v\n", md)
+	}
+	employee, err := e.repository.GetEmployee(req.ID)
+	if err != nil {
+		return nil, err
+	}
+	return &proto.EmployeeResponse{
+		Employee: entities.EmployeeFromStorageToProto(employee),
+	}, nil
+}
+
+func (e *employeeService) AddEmployee(ctx context.Context, req *proto.EmployeeRequest) (*proto.EmployeeResponse, error) {
+	req.Employee.ID = uuid.New().String()
+	employee, err := e.repository.AddEmployee(entities.EmployeeFromProtoToStorage(req.Employee))
+	if err != nil {
+		return nil, fmt.Errorf("failed to save employee: %w", err)
+	}
+	log.Printf("employee %s successfully saved;\n", employee.ID)
+	return &proto.EmployeeResponse{Employee: entities.EmployeeFromStorageToProto(employee)}, nil
+}
+func (e *employeeService) AddEmployees(stream proto.EmployeeService_AddEmployeesServer) error {
+	var savedCount int
+	var errorMessage string
+	for {
+		emp, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		emp.Employee.ID = uuid.New().String()
+		saved, err := e.repository.AddEmployee(entities.EmployeeFromProtoToStorage(emp.Employee))
+		if err != nil {
+			errorMessage += fmt.Sprintf("\n%s", err.Error())
+			continue
+		}
+		if err := stream.Send(&proto.EmployeeResponse{
+			Employee: entities.EmployeeFromStorageToProto(saved),
+		}); err != nil {
+			return err
+		}
+		savedCount++
+	}
+	if errorMessage != "" {
+		return errors.New(errorMessage)
```

Comment:
Ive also faced this issue recently. What is an error in a GRPC call? If we send an entity, but lets say the validation of one of the fields fails. Do we consider the GRPC call to have failed?

On the one hand the entity could not be persisted so the entire process failed, so you could argue that an error occurred. On the other hand, the GRPC call functioned perfectly, it has processed the entity and found there is a problem with it, so maybe we shouldnt return an error here?

Another issue is what happens if you send 1000000 entities, and one of them fails validation, is this an error? Well, its kinda a partial error right? but the GRPC call functioned perfectly for all the entities and only failed because of application-specific logic.

Ive come to the conclusion, that returning errors should be reserved for critical errors in the functioning of the GRPC stream mechanism, or perhaps where the error is so severe that the stream needs to be closed immediately, e.g. database is not available.

If the functioning of the stream is good, then I prefer to reply with explicit `Result` objects, which contain an indication of failure (and optionally of success). Including a field that has some kind of identifier for the entity, so that clients can programmatically determine which entities failed.

 https://github.com/90poe/voyage-monitor/blob/master/data-gateway/api/apiv0/apiv0.proto#L214-L216

wdyt?

Reply:
> if you send 1000000 entities, and one of them fails validation, is this an error? Well, its kinda a partial error right? ...returning errors should be reserved for critical errors
Yes, I completely agree with your comment. I think that by referring term "error" we should separate business errors and system errors. In such way it is more semantically clear, because term "critical error" is too general. There is also a concept of using specific error codes to display each of them in pretty form for user and separate from "system" errors, and in such case do better logging and alerting.